File: Back Tracking/组合/77.组合.cpp
#include <vector>

using namespace std;
/*
 * @lc app=leetcode.cn id=77 lang=cpp
 *
 * [77] 组合
 */

// @lc code=start
class Solution
{
public:
    vector<vector<int>> result;
    vector<int> path;
    void findPath(int n, int k, int startIndex)
    {
        // 数组中的元素数量达到k个，则满足条件，递归停止，并将结果添加到最后的结果集
        if (path.size() == k)
        {
            result.push_back(path);
            return;
        }

        // 这里的循环是横向遍历，将n的数组每个值一次跟后面的元素组合
        for (int i = startIndex; i <= n; i++)
        {
            // 第i个元素
            path.push_back(i);
            // 以及它后面的元素i+1，进行组合
            findPath(n, k, i + 1);
            // 组合完之后进行回溯，将i推出数组，开始统计下一个元素的组合情况
            path.pop_back();
        }
    }
    vector<vector<int>> combine(int n, int k)
    {
        if (n == 0 || k == 0)
        {
            return result;
        }

        findPath(n, k, 1);
        return result;
    }
};
// @lc code=end

File: Back Tracking/组合/77.组合.ts
/*
 * @lc app=leetcode.cn id=77 lang=typescript
 *
 * [77] 组合
 */

// @lc code=start
function combine(n: number, k: number): number[][] {
  let sum: number[][] = []

  const backTracking = (index: number, arr: number[]) => {
    if (arr.length === k) {
      // 注意这里需要使用解构重新生成一个数组
      // 避免引用修改之后影响到已经push到sum的结构
      sum.push([...arr])
      return
    }
    // i <= n的剪枝操作，k - arr.length为剩余个数
    // 如果i到n剩余不足k - arr.length个，则不需要遍历
    for (let i = index; i <= n - (k - arr.length) + 1; i++) {
      arr.push(i)
      backTracking(i + 1, arr)
      // 回溯
      arr.pop()
    }
  }

  backTracking(1, [])
  return sum
};
// @lc code=end

